Add new group page #202

SepsiLaszlo opened the pull request:
Refactor in-progress.

balintkiraly commented on app/assets/javascripts/application/suggest_user.js:
```diff
@@ -0,0 +1,10 @@
+    });
+}
+function selectUser(user_id){
+    $("#selected_user_id")[0].value=user_id
```
It would be better if we used Pure JavaScript instead of JQuery.

SepsiLaszlo replied:
Okay, I will try to rewrite it in pure JS!

falusi94 commented on app/controllers/groups_controller.rb:
```diff
@@ -29,11 +32,32 @@ def update
+  end
+
+  def create
+    group = Group.create(create_params)
+    leader = User.find(params[:selected_user_id])
+    membership = CreateMembership.call(group, leader)
+    membership.posts.where(post_type_id: PostType::DEFAULT_POST_ID).destroy_all
+    CreatePost.call(group,membership, PostType::LEADER_POST_ID)
+    redirect_to group_path(group)
```
Just as a reminder from yesterday, this logic could be encapsulated. This one should take care of:
- (pre-validation - if necessary against any condition)
- create group
- create leader membership
- notify the leader maybe?

SepsiLaszlo replied:
Thanks, I will move these to a CreateGroup service object!

falusi94 commented on app/controllers/groups_controller.rb:
```diff
@@ -29,11 +32,32 @@ def update
   end
 
+  def new
+    @parent_groups = Group.all
+                          .includes(:children)
+                          .reject { |g| g.children.empty? }
+                          .sort_by { |g| g.children.length }
+                          .reverse
```
After checking the data integrity for the resorts in the database, it can be simplified to:
```ruby
@resorts = Group.type_resort
```

SepsiLaszlo replied:
I like how simple the query became.  Should I write a rake task, for setting the resorts?

falusi94 replied:
I think it's not worth it, it is a one time job and relies on the current data in the db which is not so exact. In this case it's better and faster do a manual check.

SepsiLaszlo replied:
Should it go in like this, or we refactor it after the db integrity check?

falusi94 replied:
It can go here, first of all this is a new piece of code and I think the one you wrote depends on the database integrity too, so both way it's worth a quick check.

falusi94 commented on app/services/create_post.rb:
```diff
@@ -1,6 +1,6 @@
 class CreatePost
   def self.call(group, membership, post_type_id)
-    if post_type_id == PostType::LEADER_POST_ID
+    if (post_type_id == PostType::LEADER_POST_ID && group.leader.present?)
```
I recommend to enable `rubocop` or check it sometimes, it will help to keep the consistency in the codebase.
https://rubocop.readthedocs.io/en/latest/cops_style/#styleparenthesesaroundcondition

falusi94 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,5 +1,8 @@
 class GroupsController < ApplicationController
+  require './app/services/create_membership'
+  require './app/services/create_post'
```
Please don't use `require`, it's usually unnecessary in rails, because it uses autoloading. If you are interested in, here is some further explanation on autoloading and its behavior
https://guides.rubyonrails.org/v5.0/autoloading_and_reloading_constants.html

SepsiLaszlo replied:
You are right, I removed it. Thanks for the article, definitely going to read it!

OmTheTurtle commented on app/assets/javascripts/application/suggest_user.js:
```diff
@@ -0,0 +1,24 @@
+function suggestUser() {
+    var url = "/search/suggest_leader";
+    var params = "query=" + document.getElementsByName('group_leader_name')[0].value;
```
Let's use const here instead of var, and also string templates.

SepsiLaszlo replied:
Done!

OmTheTurtle commented on app/assets/javascripts/application/suggest_user.js:
```diff
@@ -0,0 +1,24 @@
+function suggestUser() {
+    var url = "/search/suggest_leader";
+    var params = "query=" + document.getElementsByName('group_leader_name')[0].value;
+    http = new XMLHttpRequest();
```
You can use `fetch` API, it's supported in every modern browser.

OmTheTurtle commented on app/assets/javascripts/application/suggest_user.js:
```diff
@@ -0,0 +1,24 @@
+function suggestUser() {
+    var url = "/search/suggest_leader";
+    var params = "query=" + document.getElementsByName('group_leader_name')[0].value;
+    http = new XMLHttpRequest();
```
Can you use `fetch` API here instead?

SepsiLaszlo replied:
I made an iteration with the fetch API. Is using catch and log the error a good pratice or should be removed?

OmTheTurtle commented on app/assets/javascripts/application/suggest_user.js:
```diff
@@ -0,0 +1,24 @@
+}
+
+function selectUser(user_id) {
+    $("#selected_user_id")[0].value = user_id;
```
Let's try to avoid using jQuery.

SepsiLaszlo replied:
Removed it. I think no more jQuery left in this file.

OmTheTurtle commented on app/controllers/search_controller.rb:
```diff
@@ -12,6 +14,18 @@ def suggest
+    if params.key?(:query)
+      count = Rails.configuration.x.results_per_page
+      offset = count * params[:user_count].to_i
+      user_results = SearchQuery.new.user_search(params[:query], offset, count)
+      user_results = user_results.decorate
```
Can this be written in one line?

SepsiLaszlo replied:
Fixed it. I also tried to improve the naming a little.

OmTheTurtle commented on app/views/application/_navbar.html.erb:
```diff
@@ -18,6 +18,9 @@
                 <li><%= link_to 'Küldöttek számolása', admin_delegates_count_path %></li>
                 <li><%= link_to 'Küldöttek szerkesztése', admin_delegates_path %></li>
                 <li><%= link_to 'Küldöttek export', admin_delegates_export_path %></li>
+                <%if current_user.roles.pek_admin? %>
+                  <li><%= link_to 'Kör létrehozása', new_group_path %></li>
+                <%end %>
```
Let's put a space before the `if` and `end`

OmTheTurtle commented on app/assets/javascripts/application/suggest_user.js:
```diff
@@ -0,0 +1,55 @@
+const counter = counterFactory();
+const suggestions = () => document.getElementById('user_suggestions');
+
+function suggestUser() {
```
I see in the form that this function is called every time the corresponding input changes. I'd suggest using rxjs's `debounce` to prevent too many requests.

SepsiLaszlo replied:
Okay, I will look into it.

OmTheTurtle commented on app/views/groups/new.html.erb:
```diff
@@ -0,0 +1,56 @@
+  <div class="uk-form-row">
+    <%= f.label :issvie, 'SVIE?', class: 'uk-width-1-1 uk-form-label form-label-required' do %>
+      SVIE?
+      <%#= hidden_field_tag :issvie, false %>
```
Do we need this line?

SepsiLaszlo replied:
No, I left it in the commit by mistake.

OmTheTurtle commented on test/factories/users.rb:
```diff
@@ -25,5 +25,14 @@
+
+    trait :pek_admin do
+      after(:build) do |user|
+        SystemAttribute.create(id: 2, name: 'ertekeles_idoszak', value: 'NINCSERTEKELES')
```
Do we need this here?

SepsiLaszlo replied:
I think we do. In application/_navbar.html.erb when the user is pek_admin, we run this code:
```ruby
<% unless SystemAttribute.offseason? %>
```
Which throws an error if SystemAttribute.season is not present. I'm not sure if this is a correct place to put that code, but it is necessary.

falusi94 replied:
This can be written like this
```
        create(:system_attribute_app_season)
```
though, it seems to be very extraneous to put here a system wide setting because it is used in a partial. Other option can be to put it to the test setup, which could be enough for now, but later when somebody takes a look at the test it can make them confusing, because it has nothing to do with the essence of it. The most reasonable for me to put it [here](https://github.com/kir-dev/pek-next/blob/master/spec/rails_helper.rb#L49L63) like with the semester.

OmTheTurtle commented on spec/requests/groups_controller_spec.rb:
```diff
@@ -90,4 +90,46 @@
+        get '/groups/new'
+
+        expect(response).to have_http_status :ok
+        expect(response).to render_template(:new)
```
Let's be consistent with using / not using parentheses.

falusi94 commented on spec/requests/groups_controller_spec.rb:
```diff
@@ -90,4 +90,46 @@
+      before          { login_as(pek_admin) }
+      it 'creates the group' do
+        expect{
+          post '/groups', { group: FactoryBot.attributes_for(:group),
```
I think here you can leave the `FactoryBot` as its helper is included.

Also please `headers:` and `params:` key to pass them instead of the positional args, because it's deprecated in Rails 5 and removed in 5.1.
```ruby
post '/groups', params: { group: FactoryBot.attributes_for(:group), ... }
```

falusi94 commented on app/controllers/search_controller.rb:
```diff
@@ -1,4 +1,6 @@
 class SearchController < ApplicationController
+  before_action :require_pek_admin, only: %i[suggest_leader]
```
```suggestion
  before_action :require_pek_admin, only: :suggest_leader
```
For one action you can pass an element instead of an array.

falusi94 commented on app/services/create_group.rb:
```diff
@@ -0,0 +1,9 @@
+class CreateGroup
+  def self.call(create_params, leader)
+    group = Group.create(create_params)
+    membership = CreateMembership.call(group, leader)
+    membership.posts.where(post_type_id: PostType::DEFAULT_POST_ID).destroy_all
+    CreatePost.call(group, membership, PostType::LEADER_POST_ID)
```
First of all please rebase the branch to the master, this `CreateMembership` does not exist anymore. Also instead of using `CreateMembership` (or the new `Membership::CreateService`) and deleting the default post, it seems to be much more straightforward to create the membership manually here and to add the leader post. In this case I think it's better to be less DRY, make the understanding easier, later on it can be improved and refactored when we find a fitting and reasonable way to do it.

Also other important thing, please wrap the whole method into transaction (`ActiveRecord::Base.transaction`) to avoid possible inconsistency.

SepsiLaszlo replied:
I rebased this branch to the master. I'm not sure If I understood you correctly, but i moved the leader post creation logic to the controller. I think this makes the CreateGroup service unnecessary,  since it is same as `Group.create`.

falusi94 commented on app/services/create_group.rb:
```diff
@@ -0,0 +1,9 @@
+    CreatePost.call(group, membership, PostType::LEADER_POST_ID)
+    group
+  end
+end
```
Please take care about the file endings too.

falusi94 commented on app/services/create_post.rb:
```diff
@@ -1,6 +1,6 @@
 class CreatePost
   def self.call(group, membership, post_type_id)
-    if post_type_id == PostType::LEADER_POST_ID
+    if post_type_id == PostType::LEADER_POST_ID && group.leader.present?
```
Also using this logic seems to be a bit much, you can assume that the new group has no leader.

SepsiLaszlo replied:
So the leader post for new group should not be created with `CreatePost.call`?

falusi94 replied:
I think this change is unnecessary since the `CreateGroup` does not rely on this

falusi94 replied:
I think in the end this one could be removed too.

SepsiLaszlo commented:
I refactored the CreateGroup  service object, and added notification check to the test.

Should I fix the duplication issue, spotted by codeclimate?

OmTheTurtle commented:
> Should I fix the duplication issue, spotted by codeclimate?

I thinks it's fine as-is.

falusi94 commented on app/views/search/_user_list_element.html.erb:
```diff
@@ -0,0 +1,19 @@
+<li class="uk-panel-box uk-grid uk-grid-collapse dark user-search user-result hover uk-padding-remove" onclick="window.location=''">
+  <div class="thumbnail">
+    <%= image_tag photo_path(user.screen_name), class: 'uk-thumbnail' %>
+  </div>
+  <div class="uk-padding-small next-to-thumbnail">
+    <div>
+      <%= user.link_with_compact_name class: 'uk-text-large' %>
+    </div>
+    <% Array(attributes).each do |attribute| %>
+      <div class="uk-text-break">
+        <%= user.send(attribute) %>
+        <div>
+    <% end %>
+    </div>
+    <% selectable = false unless defined?(selectable) %>
+    <% if selectable %>
+      <%= content_tag(:a, 'Kiválaszt', type: :button, class: 'uk-button uk-button-primary', onClick: "selectUser(#{user.id},'#{user.compact_name}')") %>
+    <% end %>
+  </li>
```
I'd rather not create this abstraction layer for this partial. Not optimal to duplicate code, but in this case there are slightly different expected behaviors.

For example if you take a look at this code, clicking on the `li` element in the new groups page will reload the page because of the `onclick` handler, not to mention it breaks the searching function. I'd recommend to revert the change in the `app/views/search/_suggest.html.erb` and simplify this partial view.

Also I strongly recommend avoid using `send` method if it's possible. Ruby has a great metaprogramming features, but it's better to use them in a modest way.

If you take a look at it it is much easier to understand this snippet.
```suggestion
<li class="uk-panel-box uk-grid uk-grid-collapse dark user-search user-result hover uk-padding-remove" onclick="selectUser(<%= user.id %>, '<%= user.compact_name %>')">
  <div class="thumbnail">
    <%= image_tag photo_path(user.screen_name), class: 'uk-thumbnail' %>
  </div>
  <div class="uk-padding-small next-to-thumbnail">
    <div>
      <%= user.link_with_compact_name class: 'uk-text-large' %>
    </div>
    <div class="uk-text-break">
      <%= user.room %>
    <div>
    <div class="uk-text-break">
      <%= user.email %>
    </div>
  </div>
</li>
```

SepsiLaszlo replied:
Agree, this abstraction doesn't feel right.
In my browser selecting a user doesn't reload the page, I used  `event.stopPropagation()` to prevent it.
Could not reproduce the issue with the search function. Where does it break? I'm able to search from the new group page and from /profiles/me page just fine.
l will remove send, and create a new partial for this.

falusi94 replied:
Yes, but the `event.stopPropagation()` applies only for the button, but if you click the li element it will reload it, because of this `onclick="window.location=''"`.

For the search, you miss this snippet `onclick="window.location='<%= profile_path(user.screen_name) %>'"`, so when you click on the search result for the user it does nothing.

I think this partial is fine with the suggestion and it's enough to revert the change in the suggest one.

falusi94 commented on app/assets/javascripts/application/suggest_user.js:
```diff
@@ -0,0 +1,71 @@
+    }
+}
+const counter = counterFactory();
+const suggestions = () => document.getElementById('user_suggestions');
```
Please avoid arrow functions, in the current setup it fails like most of the ES6 features.

falusi94 commented on app/services/create_group.rb:
```diff
@@ -0,0 +1,13 @@
+class CreateGroup
+  def self.call(group_params, leader_user, current_user)
+    group = nil
+    ActiveRecord::Base.transaction do
+      group             = Group.create!(group_params)
+      leader_membership = Membership.create!(group: group, user: leader_user)
+      Post.create!(membership_id: leader_membership.id, post_type_id: PostType::LEADER_POST_ID)
+
+      leader_membership.notify(:users, key: 'membership.create', notifier: current_user)
+    end
+    group
```
Just a syntactic sugar, it returns the `group` the same way. 
```suggestion
    ActiveRecord::Base.transaction do
      group             = Group.create!(group_params)
      leader_membership = Membership.create!(group: group, user: leader_user)
      Post.create!(membership_id: leader_membership.id, post_type_id: PostType::LEADER_POST_ID)

      leader_membership.notify(:users, key: 'membership.create', notifier: current_user)
      group
    end
```

SepsiLaszlo commented:
I removed the ES6 features. The assets are compiling correcting in local. Also refactored the views to be less abstract and more readable.

falusi94 commented on app/views/search/_suggest_leader.html.erb:
```diff
@@ -0,0 +1,16 @@
+<% users.each do |user| %>
+  <li class="uk-panel-box uk-grid uk-grid-collapse dark user-search user-result hover uk-padding-remove" onclick="window.location='<%= profile_path(user.screen_name) %>'">
```
```suggestion
  <li class="uk-panel-box uk-grid uk-grid-collapse dark user-search user-result hover uk-padding-remove">
```
The `onclick` handler is unnecessary here, it also makes funny things if you click on the list item instead of the button.

SepsiLaszlo commented:
I made the requested changes and rebased to master.
